sql: restore the logical order of UNION

knz opened the pull request:
In order to compute EXCEPT properly, the code for
UNION/INTERSECT/EXCEPT needs to read its right operand first to
populate the map of known values. However in the non-EXCEPT case the
resulting output is non-intuitive, as a savvy user with a reasonable
order expectation would see the right operand being consumed first,
even when not necessary.

This patch restores the "natural" order.

This is not a user-facing change worth noting in docs: the ordering of
UNION's result has never been guaranteed without an ORDER BY clause.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/18571)
<!-- Reviewable:end -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> thanks buddy!

Please take the joy of removing this TODO:
https://github.com/cockroachdb/cockroach/blob/b192b3701fe98d02a48b6e03cf0bf5a565d79e40/pkg/sql/txn_restart_test.go#L1801

---

Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/plan_columns.go, line 114 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18571#-KuMD0kfRxunS92PYy4P:-KuMD0kfRxunS92PYy4Q:b-d4dkal) ([raw file](https://github.com/cockroachdb/cockroach/blob/727dde54946f6622502b0638a0cae01e82c79a6e/pkg/sql/plan_columns.go#L114)):*
> ```Go
> 		return getPlanColumns(n.plan, mut)
> 	case *unionNode:
> 		if n.inverted {
> ```

is this change necessary? The set of columns should be the same in the two sources, no?

---

*[pkg/sql/union.go, line 146 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18571#-KuMC1VgBSLf6PyMIggH:-KuMC1VgBSLf6PyMIggI:b-89vw6m) ([raw file](https://github.com/cockroachdb/cockroach/blob/727dde54946f6622502b0638a0cae01e82c79a6e/pkg/sql/union.go#L146)):*
> ```Go
> 	emit        unionNodeEmit
> 	scratch     []byte
> 	inverted    bool
> ```

please comment the field :)

And probably a note about it / about the ordering of results should be put in the node's description too.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18571#-:-KuMCqTRabTi9qM7Jcdh:bmn4pwa)*
<!-- Sent from Reviewable.io -->

knz commented:
Done. TFYR!

---

Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/plan_columns.go, line 114 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18571#-KuMD0kfRxunS92PYy4P:-KuMJphp9-_uwml3BPMJ:bsxfbvi) ([raw file](https://github.com/cockroachdb/cockroach/blob/727dde54946f6622502b0638a0cae01e82c79a6e/pkg/sql/plan_columns.go#L114)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

is this change necessary? The set of columns should be the same in the two sources, no?
</blockquote></details>

No: SELECT * FROM kv UNION VALUES(...)
The VALUES clause has no useful column names. This is being tested.

---

*[pkg/sql/union.go, line 146 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18571#-KuMC1VgBSLf6PyMIggH:-KuMKj91JQlIAEKy1CIB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/727dde54946f6622502b0638a0cae01e82c79a6e/pkg/sql/union.go#L146)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please comment the field :)

And probably a note about it / about the ordering of results should be put in the node's description too.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18571#-:-KuMJoCIDuBZzPoijLtj:b-qqwx6r)*
<!-- Sent from Reviewable.io -->

knz committed:
sql: restore the logical order of UNION

In order to compute EXCEPT properly, the code for
UNION/INTERSECT/EXCEPT needs to read its right operand first to
populate the map of known values. However in the non-EXCEPT case the
resulting output is non-intuitive, as a savvy user with a reasonable
order expectation would see the right operand being consumed first,
even when not necessary.

This patch restores the "natural" order.

This is not a user-facing change worth noting in docs: the ordering of
UNION's result has never been guaranteed without an ORDER BY clause.
